envision/server.py

```diff
@@ -195,13 +195,13 @@ async def run_loop():
                     frames_to_send = [frame_ptr]
                     self._seek = None
 
-                if len(frames_to_send) > 0:
-                    closed = self._push_frames_to_web_client(frames_to_send)
-                    if closed:
-                        self._log.debug("Socket closed, exiting")
-                        return
+                assert len(frames_to_send) > 0
```

Comment:
When will len(frames_to_send) <=0?

`self._wait_for_next_frame(frame_ptr)` appear to only return when `len(frames_to_send) > 0`. Thus if needed, the `len(frames_to_send)` can be checked inside the above `if ...` block to avoid assertion in every while loop.

Reply:
@Gamenot The assertion isn't caught.  In general, I don't think asserts should be used for errors that a user will experience, nor should / need to be caught.  I use them for messages to _coders_ (i.e., us) to indicate some assumption has been violated (and simultaneously for documentation).  This assert isn't to protect from a possible runtime error, but rather to warn any coder who makes changes in the future.  Thus, I think an immediate crash is an appropriate (indeed, the correct) outcome.

@Adaickalavan It should never be 0.  (that's why I added the assert and removed the `if`.)

But I think having the assert in the while loop (outside of the `if`) is useful because it is asserting that, under no circumstances will `_push_frames_to_web_client()` be called without frames to send.
This way, if the loop here ever gets made more complex, it will be clear that we're always expecting to have frames to send.  (also, I think the overhead of an assert is pretty low.)